pmkoalas/models/petrinet.py

```diff
@@ -0,0 +1,341 @@
+import xml.etree.ElementTree as ET
+
+ENCODING='unicode'
+
+def steq(self,other):
+    if type(other) is type(self):
+        return self.__dict__ == other.__dict__
+    return False
+
+
+
+class LabelledPetriNet:
+    def __init__(self,places,transitions,arcs,label=None):
+        self._places = places
+        self._transitions = transitions
+        self._arcs = arcs
+        self._label = label
+
+    @property 
+    def places(self):
+        return self._places
+
+    @property
+    def transitions(self):
+        return self._transitions
+
+    @property
+    def arcs(self):
+        return self._arcs
+
+    @property
+    def label(self):
+        return self._label
+
+    def __eq__(self,other):
+        if isinstance(other,self.__class__):
+            return self._label  == other._label and \
+                   self._places == other._places and \
+                   self._transitions == other._transitions and \
+                   self._arcs   == self._arcs
+        return False
+
+    def reprcontents(self):
+        return f"places: {self._places} transitions: {self._transitions} arcs: {self._arcs}"
+
+    def __repr__(self):
+        return f"LabelledPetriNet:{self._label} " + self.reprcontents()
+
+    def __str__(self):
+        return "LabelledPetriNet(" + self._label + ")\n" \
+            +  " Places: " + str(self.places) + "\n" \
+            +  " Trans: " + str(self.transitions) + "\n" \
+            +  " Arcs: " + str(self.arcs)
+    
+
+class MutableLabelledPetriNet(LabelledPetriNet):
+    def __init__(self,label=None):
+        super(MutableLabelledPetriNet,self).__init__(set(),set(),set(),label)
+
+    def addPlace(self,place):
+        self._places.add(place)
+
+    def addTransition(self,tran):
+        self._transitions.add(tran)
+
+    def addArc(self,arc):
+        self._arcs.add(arc)
+
+    def addArcForNodes(self,fromNode,toNode):
+        self.addArc(Arc(fromNode,toNode))
+
+    def __repr__(self):
+        return f"MutableLabelledPetriNet:{self._label} " + self.reprcontents()
+
+    def __eq__(self,other):
+        if isinstance(other,LabelledPetriNet):
+            return self._label  == other._label and \
+                   self._places == other._places and \
+                   self._transitions == other._transitions and \
+                   self._arcs   == self._arcs
+        return False
+
+def verbosecmp(obj1,obj2):
+    if obj1 == obj2:
+        return "Same"
+    result = ""
+    if obj1.__dict__.keys() != obj2.__dict__.keys():
+        ks1 = set(obj1.__dict__.keys())
+        ks2 = set(obj2.__dict__.keys())
+        result += f"Attributes not in both objects: {ks1 ^ ks2}\n"
+    if obj1.__dict__ != obj2.__dict__:
+        for k in obj1.__dict__:
+            if k in obj2.__dict__ and obj1.__dict__[k] != obj2.__dict__[k] :
+                result += f"Attribute {k} differs:\n"\
+                        + f"  {obj1.__dict__[k]} vs\n  {obj2.__dict__[k]}\n"
+    return result
+
+
+
+class Place:
```

Comment:
Why are these templates not used in typing for the structure? I can see why you talked about have a more modular structure for the net and its components. You have already added a lot of "extras" for the simple components of a petri net. I do agree with the need for a unique identifier, but perhaps we should handle that first and leave the option to set it off by default.

Reply:
By template do you mean class? Do you mean in function signatures? It's mainly because I have trouble taking non-enforced typing seriously.

Reply:
These classes do not have any functional behaviours other than to be holding data, I called these templates. Technically these can be dramatically reduced using dataclasses which will do all the work for you.

Type hinting allows for smoother construction, many will read at least the source for your function call. By having the hints there, a user can quicky find the appropriate class to use. The whole module already uses this pattern.

see https://docs.python.org/3/library/dataclasses.html

Reply:
I can make you a decorator to enforce hints if you want, its not hard.

Reply:
Templates is usually associated with generics because that is the jargon in C++.

These classes are likely to grow extra behaviour. I have used dataclasses in other places, but here they would indicate a design intent to keep these as simply records.

Runtime decorators to enforce types aren't high value.

I'm not against type hinting, it's fine, I will add it. But it is a bit silly without build-time tooling.